controllers/cloudstackmachine_controller.go

```diff
@@ -127,61 +133,25 @@ func (r *CloudStackMachineReconciliationRunner) ConsiderAffinity() (ctrl.Result,
 
 // SetFailureDomainOnCSMachine sets the failure domain the machine should launch in.
 func (r *CloudStackMachineReconciliationRunner) SetFailureDomainOnCSMachine() (retRes ctrl.Result, reterr error) {
-	// Set ZoneID on csMachine.
-	if util.IsControlPlaneMachine(r.CAPIMachine) { // Use failure domain zone.
-		r.ReconciliationSubject.Status.ZoneID = *r.CAPIMachine.Spec.FailureDomain
-	} else { // Specified by Machine Template or Random zone.
-		if r.ReconciliationSubject.Spec.ZoneID != "" {
-			if zone, foundZone := r.CSCluster.Status.Zones[r.ReconciliationSubject.Spec.ZoneID]; foundZone { // ZoneID Specified.
-				r.ReconciliationSubject.Status.ZoneID = zone.ID
-			} else {
-				return ctrl.Result{}, errors.Errorf("could not find zone by zoneID: %s", r.ReconciliationSubject.Spec.ZoneID)
-			}
-		} else if r.ReconciliationSubject.Spec.ZoneName != "" {
-			for _, zone := range r.Zones.Items {
-				if zone.Spec.Name == r.ReconciliationSubject.Spec.ZoneName {
-					r.ReconciliationSubject.Status.ZoneID = zone.Spec.ID
-					break
-				}
-			}
-			if r.ReconciliationSubject.Status.ZoneID == "" {
-				return ctrl.Result{}, errors.Errorf("could not find zone by zoneName: %s", r.ReconciliationSubject.Spec.ZoneName)
-			}
-		} else { // No Zone Specified, pick a Random Zone.
-			// TODO: Add an additional controller test for this once controller test upgrades are merged.
-			if len(r.Zones.Items) < 1 { // Double check that zones are present.
-				return r.RequeueWithMessage("no zones found, requeueing")
-			}
-			randNum := (rand.Int() % len(r.Zones.Items)) // #nosec G404 -- weak crypt rand doesn't matter here.
-			r.ReconciliationSubject.Status.ZoneID = r.Zones.Items[randNum].Spec.ID
-		}
-	}
-	for idx, zone := range r.Zones.Items {
-		if zone.Spec.ID == r.ReconciliationSubject.Status.ZoneID {
-			r.FailureDomain = &r.Zones.Items[idx]
-			break
+	if r.ReconciliationSubject.Spec.FailureDomainName == "" { // Needs random FD, but not yet set.
+		if util.IsControlPlaneMachine(r.CAPIMachine) { // Is control plane machine -- CAPI will specify.
```

Comment:
```suggestion
		r.CAPIMachine.Spec.FailureDomain != nil and r.CAPIMachine.Spec.FailureDomain != ""
```
As far as I know the etcdadm controller also sets the CAPIMachine.Spec.FailureDomain.

Reply:
Thanks, updated.